modules/afmongodb/afmongodb-worker.c

```diff
@@ -388,17 +482,38 @@ _worker_insert(LogThreadedDestWorker *s, LogMessage *msg)
             evt_tag_value_pairs("message", owner->vp, msg, &options),
             evt_tag_str("driver", owner->super.super.super.id));
 
-
   if (!owner->collection_is_literal_string)
     {
+      g_string_assign(self->last_collection, self->collection->str);
       const gchar *new_collection = _format_collection_template(self, msg);
-      if (!_switch_collection(self, new_collection))
+      bool should_switch_collection = (strcmp(self->last_collection->str, new_collection) != 0);
+
+      if (should_switch_collection && !_switch_collection(self, new_collection))
         return LTR_ERROR;
     }
 
   bson_error_t error;
-  success = mongoc_collection_insert(self->coll_obj, MONGOC_INSERT_NONE,
-                                     (const bson_t *)self->bson, NULL, &error);
+  if (owner->use_bulk)
+    {
+      if (self->bulk_op == NULL)
+        self->bulk_op = mongoc_collection_create_bulk_operation_with_opts(self->coll_obj, self->bson_opts);
+      if (self->bulk_op)
+        mongoc_bulk_operation_set_bypass_document_validation (self->bulk_op, owner->bulk_bypass_validation);
+
+      if (self->bulk_op == NULL)
+        {
+          msg_error("Failed to create MongoDB bulk operation",
+                    evt_tag_int("time_reopen", self->super.time_reopen),
+                    evt_tag_str("driver", owner->super.super.super.id));
+          return LTR_ERROR;
+        }
+
+      mongoc_bulk_operation_insert(self->bulk_op, (const bson_t *)self->bson);
+      return LTR_QUEUED;
```

Comment:
Optional: I'd extract this to a function so that `_worker_insert()` could look like this:

```
if (owner->use_bulk)
   _build_insert();
else
  mongoc_collection_insert();

```

Reply:
good point